src/Castle.Windsor/MicroKernel/Registration/BasedOnDescriptor.cs

```diff
@@ -535,11 +536,14 @@ private static bool IsBasedOnGenericInterface(Type type, Type basedOn, out Type[
 			var types = new List<Type>(4);
 			foreach (var @interface in type.GetInterfaces())
 			{
-				if (@interface.IsGenericType &&
+				if (@interface.GetTypeInfo().IsGenericType &&
 				    @interface.GetGenericTypeDefinition() == basedOn)
 				{
-					if (@interface.ReflectedType == null &&
-					    @interface.ContainsGenericParameters)
+					if (@interface.GetTypeInfo().ContainsGenericParameters
+#if FEATURE_REFLECTION_REFLECTEDTYPE
+						&& @interface.ReflectedType == null
+#endif
```

Comment:
I'm not a fan of the two usages of `FEATURE_REFLECTION_REFLECTEDTYPE`, this isn't conditionally compiling functionality out (e.g. serialization or System.Web support) but something that shouldn't really differ between the builds on both runtimes otherwise it could introduce subtle bugs.

I don't know if DeclaringType could be used in place of ReflectedType here, but we removed the one usage of ReflectedType from Castle Core in https://github.com/castleproject/Core/commit/87c36c31490be1a7638ce4b0fb5441b5c4854c89 by passing the object's type around.